### ISSUE TITLE
2016-02-25-160900_remove_null_consumer_id: [cassandra error] failed waiting for schema consensus: timeout

### ISSUE THREAD

**Issue by harf18** [Kong/kong#3005]
When I do `kong migrations up --v` after configuring a single instance of Kong to use a Cassandra cluster (with 3 nodes), I've got the error 

> kong migrations up --v
2017/11/02 12:02:19 [verbose] Kong: 0.11.0
2017/11/02 12:02:19 [verbose] reading config file at /etc/kong/kong.conf
2017/11/02 12:02:19 [verbose] prefix in use: /usr/local/kong
2017/11/02 12:02:19 [verbose] running datastore migrations
2017/11/02 12:02:19 [info] migrating core for keyspace kong
2017/11/02 12:02:21 [info] core migrated up to: 2015-01-12-175310_skeleton
2017/11/02 12:02:39 [info] core migrated up to: 2015-01-12-175310_init_schema
2017/11/02 12:02:39 [info] core migrated up to: 2015-11-23-817313_nodes
Error:
/usr/local/share/lua/5.1/kong/cmd/migrations.lua:34: [cassandra error] Error during migration 2016-02-25-160900_remove_null_consumer_id: [cassandra error] failed waiting for schema consensus: timeout
stack traceback:
        [C]: in function 'assert'
        /usr/local/share/lua/5.1/kong/cmd/migrations.lua:34: in function 'cmd_exec'
        /usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:88>
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:45>
        /usr/local/bin/kong:7: in function 'file_gen'
        init_worker_by_lua:39: in function <init_worker_by_lua:37>
        [C]: in function 'xpcall'
        init_worker_by_lua:46: in function <init_worker_by_lua:44>

### Additional Details & Logs

- Kong version : 0.11.0
- Kong configuration

>#------------------------------------------------------------------------------
#DATASTORE
#------------------------------------------------------------------------------
database = cassandra             # Determines which of PostgreSQL or Cassandra
cassandra_contact_points = 10.253.40.11,10.253.40.12,10.253.40.13  # A comma-separated list of contact
cassandra_port = 9042           # The port on which your nodes are listening
cassandra_keyspace = kong       # The keyspace to use in your cluster.
cassandra_timeout = 5000        # Defines the timeout (in ms), for reading
#cassandra_ssl = off             # Toggles client-to-node TLS connections
#cassandra_ssl_verify = off      # Toggles server certificate verification if
#cassandra_username = kong       # Username when using the
#cassandra_password =            # Password when using the
#cassandra_consistency = ONE   # Consistency setting to use when reading/
#cassandra_lb_policy = RoundRobin  # Load balancing policy to use when
#cassandra_local_datacenter =    # When using the `DCAwareRoundRobin` load
#cassandra_repl_strategy = SimpleStrategy  # When migrating for the first time,
#cassandra_repl_factor = 1       # When migrating for the first time, Kong
#cassandra_data_centers = dc1:2,dc2:3  # When migrating for the first time,
cassandra_schema_consensus_timeout = 6000  # Defines the timeout (in ms) for
#------------------------------------------------------------------------------
#DATASTORE CACHE
#------------------------------------------------------------------------------
#db_update_frequency = 5         # Frequency (in seconds) at which to check for
db_update_propagation = 5       # Time (in seconds) taken for an entity in the
#db_cache_ttl = 3600             # Time-to-live (in seconds) of an entity from
                                 # the datastore when cached by this node.
                                 # Database misses (no entity) are also cached
                                 # according to this setting.
                                 # If set to 0, such cached entities/misses
                                 # never expire.

- Operating System : Linux 4.9.0-3-amd64 #1 SMP Debian 4.9.30-2+deb9u3 (2017-08-06) x86_64 GNU/Linux

- Kong migration (i try also with  cassandra_schema_consensus_timeout  set to 60000) :
> kong migrations up --vv
2017/11/02 11:51:31 [verbose] Kong: 0.11.0
2017/11/02 11:51:31 [debug] ngx_lua: 10008
2017/11/02 11:51:31 [debug] nginx: 1011002
2017/11/02 11:51:31 [debug] Lua: LuaJIT 2.1.0-beta2
2017/11/02 11:51:31 [verbose] reading config file at /etc/kong/kong.conf
2017/11/02 11:51:31 [warn] You are using Cassandra but your 'db_update_propagation' setting is set to '0' (default). Due to the distributed nature of Cassandra, you should increase this value.
2017/11/02 11:51:31 [debug] admin_access_log = "logs/admin_access.log"
2017/11/02 11:51:31 [debug] admin_error_log = "logs/error.log"
2017/11/02 11:51:31 [debug] admin_http2 = false
2017/11/02 11:51:31 [debug] admin_listen = "0.0.0.0:8001"
2017/11/02 11:51:31 [debug] admin_listen_ssl = "0.0.0.0:8444"
2017/11/02 11:51:31 [debug] admin_ssl = true
2017/11/02 11:51:31 [debug] anonymous_reports = true
2017/11/02 11:51:31 [debug] cassandra_consistency = "ONE"
2017/11/02 11:51:31 [debug] cassandra_contact_points = {"10.253.40.11","10.253.40.12","10.253.40.13"}
2017/11/02 11:51:31 [debug] cassandra_data_centers = {"dc1:2","dc2:3"}
2017/11/02 11:51:31 [debug] cassandra_keyspace = "kong"
2017/11/02 11:51:31 [debug] cassandra_lb_policy = "RoundRobin"
2017/11/02 11:51:31 [debug] cassandra_port = 9042
2017/11/02 11:51:31 [debug] cassandra_repl_factor = 1
2017/11/02 11:51:31 [debug] cassandra_repl_strategy = "SimpleStrategy"
2017/11/02 11:51:31 [debug] cassandra_schema_consensus_timeout = 6000
2017/11/02 11:51:31 [debug] cassandra_ssl = false
2017/11/02 11:51:31 [debug] cassandra_ssl_verify = false
2017/11/02 11:51:31 [debug] cassandra_timeout = 5000
2017/11/02 11:51:31 [debug] cassandra_username = "kong"
2017/11/02 11:51:31 [debug] client_body_buffer_size = "8k"
2017/11/02 11:51:31 [debug] client_max_body_size = "0"
2017/11/02 11:51:31 [debug] client_ssl = false
2017/11/02 11:51:31 [debug] custom_plugins = {}
2017/11/02 11:51:31 [debug] database = "cassandra"
2017/11/02 11:51:31 [debug] db_cache_ttl = 3600
2017/11/02 11:51:31 [debug] db_update_frequency = 5
2017/11/02 11:51:31 [debug] db_update_propagation = 0
2017/11/02 11:51:31 [debug] dns_error_ttl = 1
2017/11/02 11:51:31 [debug] dns_hostsfile = "/etc/hosts"
2017/11/02 11:51:31 [debug] dns_no_sync = false
2017/11/02 11:51:31 [debug] dns_not_found_ttl = 30
2017/11/02 11:51:31 [debug] dns_order = {"LAST","SRV","A","CNAME"}
2017/11/02 11:51:31 [debug] dns_resolver = {}
2017/11/02 11:51:31 [debug] dns_stale_ttl = 4
2017/11/02 11:51:31 [debug] error_default_type = "text/plain"
2017/11/02 11:51:31 [debug] http2 = false
2017/11/02 11:51:31 [debug] latency_tokens = true
2017/11/02 11:51:31 [debug] log_level = "notice"
2017/11/02 11:51:31 [debug] lua_code_cache = "on"
2017/11/02 11:51:31 [debug] lua_package_cpath = ""
2017/11/02 11:51:31 [debug] lua_package_path = "?/init.lua;./kong/?.lua"
2017/11/02 11:51:31 [debug] lua_socket_pool_size = 30
2017/11/02 11:51:31 [debug] lua_ssl_verify_depth = 1
2017/11/02 11:51:31 [debug] mem_cache_size = "128m"
2017/11/02 11:51:31 [debug] nginx_daemon = "on"
2017/11/02 11:51:31 [debug] nginx_optimizations = true
2017/11/02 11:51:31 [debug] nginx_user = "nobody nobody"
2017/11/02 11:51:31 [debug] nginx_worker_processes = "auto"
2017/11/02 11:51:31 [debug] pg_database = "kong"
2017/11/02 11:51:31 [debug] pg_host = "127.0.0.1"
2017/11/02 11:51:31 [debug] pg_port = 5432
2017/11/02 11:51:31 [debug] pg_ssl = false
2017/11/02 11:51:31 [debug] pg_ssl_verify = false
2017/11/02 11:51:31 [debug] pg_user = "kong"
2017/11/02 11:51:31 [debug] prefix = "/usr/local/kong/"
2017/11/02 11:51:31 [debug] proxy_access_log = "logs/access.log"
2017/11/02 11:51:31 [debug] proxy_error_log = "logs/error.log"
2017/11/02 11:51:31 [debug] proxy_listen = "0.0.0.0:80"
2017/11/02 11:51:31 [debug] proxy_listen_ssl = "0.0.0.0:443"
2017/11/02 11:51:31 [debug] real_ip_header = "X-Real-IP"
2017/11/02 11:51:31 [debug] real_ip_recursive = "off"
2017/11/02 11:51:31 [debug] server_tokens = true
2017/11/02 11:51:31 [debug] ssl = true
2017/11/02 11:51:31 [debug] ssl_cipher_suite = "modern"
2017/11/02 11:51:31 [debug] ssl_ciphers = "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256"
2017/11/02 11:51:31 [debug] trusted_ips = {}
2017/11/02 11:51:31 [debug] upstream_keepalive = 60
2017/11/02 11:51:31 [verbose] prefix in use: /usr/local/kong
2017/11/02 11:51:31 [verbose] running datastore migrations
2017/11/02 11:51:31 [info] migrating core for keyspace kong
Error:
/usr/local/share/lua/5.1/kong/cmd/migrations.lua:34: [cassandra error] Error during migration 2016-02-25-160900_remove_null_consumer_id: [cassandra error] failed waiting for schema consensus: timeout
stack traceback:
        [C]: in function 'assert'
        /usr/local/share/lua/5.1/kong/cmd/migrations.lua:34: in function 'cmd_exec'
        /usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:88>
        [C]: in function 'xpcall'
        /usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:45>
        /usr/local/bin/kong:7: in function 'file_gen'
        init_worker_by_lua:39: in function <init_worker_by_lua:37>
        [C]: in function 'xpcall'
        init_worker_by_lua:46: in function <init_worker_by_lua:44>

**Comment by thibaultcha** [comment]
Hi,
Thanks for the report. Since I have been seeing this particular error in that particular migration making some appearances for various users at this point, I'll have a look and try to make some sense of it...

I would also recommend that on your side, you try to make sure your cluster, and especially your peer-to-peer communications are healthy. I seem to recall having encountered this issue when a node is lagging behind due to some overloading, or unhealthy gossiping with other nodes, and never receiving schema updates, thus rightfully triggering this error.

BTW: it seems like you set your consensus timeout to `6000` (6s) in your second test, and not `60000` (60s) as you said you intended to. I recall having seen some users encountering the same issue successfully run this migration once bumping this value to 60s. You might want to try again.

**Comment by harf18** [comment]
Thanks for reply.
for the consenus timeout, i wasn't clear : I try 6000, 60000 (even 3600000 the morning...).

I try something else : I reset all my cassandra nodes with :

> sudo rm -rf /var/lib/cassandra/commitlog/*
> sudo rm -rf /var/lib/cassandra/data/*

I start only one seed node.

I execute the migrations command on kong, after I can start kong

> 2017/11/03 16:03:15 [verbose] Kong: 0.11.0
2017/11/03 16:03:15 [verbose] reading config file at /etc/kong/kong.conf
2017/11/03 16:03:15 [verbose] prefix in use: /usr/local/kong
2017/11/03 16:03:15 [verbose] preparing nginx prefix directory at /usr/local/kong
2017/11/03 16:03:15 [verbose] SSL enabled, no custom certificate set: using default certificate
2017/11/03 16:03:15 [verbose] default SSL certificate found at /usr/local/kong/ssl/kong-default.crt
2017/11/03 16:03:15 [verbose] Admin SSL enabled, no custom certificate set: using default certificate
2017/11/03 16:03:15 [verbose] admin SSL certificate found at /usr/local/kong/ssl/admin-kong-default.crt
2017/11/03 16:03:15 [warn] ulimit is currently set to "1024". For better performance set it to at least "4096" using "ulimit -n"
2017/11/03 16:03:15 [verbose] checking Cassandra schema consensus...
2017/11/03 16:03:15 [verbose] Cassandra schema consensus: reached
2017/11/03 16:03:15 [info] Kong started

I start a new node and wait it start...

I do a kong restart, but i have the same problem...

> 2017/11/03 16:06:18 [verbose] Kong: 0.11.0
2017/11/03 16:06:18 [verbose] reading config file at /etc/kong/kong.conf
2017/11/03 16:06:18 [verbose] prefix in use: /usr/local/kong
2017/11/03 16:06:18 [verbose] preparing nginx prefix directory at /usr/local/kong
2017/11/03 16:06:18 [verbose] SSL enabled, no custom certificate set: using default certificate
2017/11/03 16:06:18 [verbose] default SSL certificate found at /usr/local/kong/ssl/kong-default.crt
2017/11/03 16:06:18 [verbose] Admin SSL enabled, no custom certificate set: using default certificate
2017/11/03 16:06:18 [verbose] admin SSL certificate found at /usr/local/kong/ssl/admin-kong-default.crt
2017/11/03 16:06:18 [warn] ulimit is currently set to "1024". For better performance set it to at least "4096" using "ulimit -n"
2017/11/03 16:06:18 [verbose] checking Cassandra schema consensus...
2017/11/03 16:06:18 [verbose] Cassandra schema consensus: not reached
2017/11/03 16:06:18 [verbose] could not start Kong, stopping services
2017/11/03 16:06:18 [verbose] stopped services
Error:
/usr/local/share/lua/5.1/kong/cmd/start.lua:62: /usr/local/share/lua/5.1/kong/cmd/start.lua:48: Cassandra has not reached cluster consensus yet

My Cassandra configuration 

> cluster_name: 'kong_cluster'
> num_tokens: 256
> seed_provider:
>    \- class_name: org.apache.cassandra.locator.SimpleSeedProvider
       parameters:
>          \- seeds: "10.253.40.11"
listen_address: 10.253.40.11
rpc_address: 10.253.40.11
endpoint_snitch: GossipingPropertyFileSnitch

**Comment by harf18** [comment]
Another information 
I do a `sudo nodetool describecluster` and I have 2 distinct shema versions but i don't understand why !

> Cluster Information:
        Name: kong_cluster
        Snitch: org.apache.cassandra.locator.DynamicEndpointSnitch
        Partitioner: org.apache.cassandra.dht.Murmur3Partitioner
        Schema versions:
                491bbcb7-09dc-3fed-b776-f7de91bfc500: [10.253.40.11]
                53eee78d-b3d0-3795-b2d6-c943725632e8: [10.253.40.13, 10.253.40.12]

**Comment by jeremyjpj0916** [comment]
Are you me :) ? I was running into this same exact situation. I am not sure the environment in which you are running but I decided to use Kong's Kubernetes cassandra migration yaml to initiate the Kong schema dump into a fresh keystore and had similar findings as yourself. No matter what schema consensus timeout occurs I get a error when running in debug as:

`2017/11/06 20:34:52 [info] migrating core for keyspace kong
2017/11/06 20:34:52 [info] core migrated up to: 2015-01-12-175310_skeleton
2017/11/06 20:34:52 [info] core migrated up to: 2015-01-12-175310_init_schema
2017/11/06 20:34:53 [info] core migrated up to: 2015-11-23-817313_nodes
2017/11/06 20:34:54 [info] core migrated up to: 2016-02-25-160900_remove_null_consumer_id
Error: 
/usr/local/share/lua/5.1/kong/cmd/migrations.lua:34: [cassandra error] Error during migration 2017-01-24-132600_upstream_timeouts_2: [Read timeout] Operation timed out - received only 0 responses.
stack traceback:
    [C]: in function 'assert'
    /usr/local/share/lua/5.1/kong/cmd/migrations.lua:34: in function 'cmd_exec'
    /usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:88>
    [C]: in function 'xpcall'
    /usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:45>
    /usr/local/bin/kong:7: in function 'file_gen'
    init_worker_by_lua:39: in function <init_worker_by_lua:37>
    [C]: in function 'xpcall'
    init_worker_by_lua:46: in function <init_worker_by_lua:44>
2017/11/06 20:34:54 [info] core migrated up to: 2016-02-29-121813_remove_ttls
2017/11/06 20:34:54 [info] core migrated up to: 2016-09-05-212515_retries_step_1`

A few things I am currently looking into on my end is I believe all my nodes are "seeds" and I am wondering if that could partially be the culprit. From researching the error for me "Operation timed out - received only 0 responses." Means that Cassandra was unable to replicate in time across to a given Cassandra node from the primary Cassandra node your Kong node was calling. But the fact no matter what the timeout gets set to for the consensus timeout it happens anyways is weird, seems like that config value does not really impact the Kong schema performance timeout (even at 60,000) still fails at similar progression steps during fresh migrations. I will say that after I run the same migrate job 3 times, by the 3rd time the migrations will have completed successfully, so this is not a show stopper to me testing Kong in a dev environment but certainly a concerning nuisance. Will report back if I find anything else on the matter. I am running a 5 node Cassandra cluster in 1 DC doing a simple strategy round robin currently with Kong set to ONE on r/w. Cassandra replication set to 2 currently(down from 5 thinking that was the issue but still did not fix it). Note after running sudo nodetool describecluster I only see 1 schema version for all 5 node IP's unlike yours above. The above tells me your node schema's are out of sync.

**Comment by gdrouet** [comment]
Same problem here. Do you have any update on this?

**Comment by CAFontana** [comment]
I've run into this same issue running a cluster of 3 with all of them as seeds.  Running `kong migrations up` 2+ times also resolves this for me but is a nuisance since we deploy through ansible scripts which will fail.

**Comment by jeremyjpj0916** [comment]
@gdrouet @CAFontana I have not found a fix yet, its not a killer to me because I just run the migrations multiple times and it works a little more each time until the last one does a large majority of the migrations, modifying the kong config timeouts related to the db does not alleviate the issue. I am starting to not think its our C* clusters but maybe somewhere in the migrations process that times out and does not have a configurable timeout.. hard to say. Will update this topic if I ever find the bugger.

**Comment by gdrouet** [comment]
I'm also using Ansible and it's hard for ops to hear that the playbook needs to be run several times until it works. In my case I don't need the migration and I just need to initialize the schema from scratch. Do you think it's possible to dump the 0.12 schema and restore it instead of running the migrations? This would be a nice temporary workaround in my case.
Thanks a lot for your feedbacks!

**Comment by pamiel** [comment]
Same issue encountered on my side (on a 0.12.3):
```
migrating core for keyspace api_gateway
core migrated up to: 2015-01-12-175310_skeleton
core migrated up to: 2015-01-12-175310_init_schema
core migrated up to: 2015-11-23-817313_nodes
Error: [cassandra error] Error during migration 2016-02-25-160900_remove_null_consumer_id: [cassandra error] failed waiting for schema consensus: timeout
```

Did anyone make any progress in the analysis of the root cause of the issue (last post was in Jan)?
Any solution or workaround... else than just trying 2 or 3 times the migration... as to be honest this is not a very predictive solution, indeed?

Thanks for your help!

**Comment by rahimimo** [comment]
Root cause is not known to me, I run a script to continue migration till success.

**Comment by pamiel** [comment]
Indeed, I think we are talking here about 2 different problems (sorry for that, I participated to the confusion!):
- Initial problem from @harf18, on `2016-02-25-160900_remove_null_consumer_id`: on my side, it looks this step of the migration is taking quite a long time. So I solved the issue by increasing the `cassandra_schema_consensus_timeout` parameter to 2 minutes (=120000ms).
- But then appeared the second issue mentioned by @jeremyjpj0916 on `2017-01-24-132600_upstream_timeouts_2`. This second issue was, for me, initially hidden by the first problem. This issue is indeed logged here https://github.com/Kong/kong/issues/3392, and I'm about to submit a PR for the solution: replacing the raw `SELECT * FROM apis` (that does not wait schema consensus nor performs retry on read failure) by the `dao.apis:find_all()` command using the DAO layer (that performs both schema consensus wait and retry on read failure).

With these 2 updates, both issues are gone on my side, and no need to run and run the migration again until it succeeds...
